### PR TITLE
[docs] Docs for ColossalaiStrategy

### PR DESCRIPTION
**Pull request by 1SAA**
## What does this PR do?

Fixes some RuntimeError in the example [repository](https://github.com/hpcaitech/ColossalAI-Pytorch-lightning).

And adds docs for newly merged [ColossalaiStrategy](https://github.com/Lightning-AI/lightning/pull/14224).

## Before submitting

- [ ] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [ ] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [ ] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 1]
```diff
@@ -55,6 +55,94 @@ Sharding techniques help when model sizes are fairly large; roughly 500M+ parame
...
+        It is recomended to use ``HybridAdam``, since it updates parameters in GPU and CPU both.
+
+    *   Your model must be created in the function ``LightningModule.configure_sharded_model()``.
+
+    *   ``ColossalaiStrategy`` doesn't support gradient accumulation now.
+
+.. _colossal_placement_policy:
+
+Placement Policy
+================================
```
Comment (rohitgr7) [inline docs/source-pytorch/advanced/model_parallel.rst:118]:
```suggestion
Placement Policy
================
```

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 2]
```diff
@@ -55,6 +55,94 @@ Sharding techniques help when model sizes are fairly large; roughly 500M+ parame
...
+
+    *   The only accelerator which ColossalAI supports is ``"gpu"``. But CPU resources will be used when the placement policy is set to "auto" or "cpu".
+
+    *   The only precision which ColossalAI allows is 16 (FP16).
+
+    *   It only supports a single optimizer, which must be ``colossalai.nn.optimizer.CPUAdam`` or ``colossalai.nn.optimizer.
+        HybridAdam`` now. You can set ``adamw_mode`` to False to use normal Adam. Noticing that ``HybridAdam`` is highly optimized, it uses fused CUDA kernel and parallel CPU kernel.
+        It is recomended to use ``HybridAdam``, since it updates parameters in GPU and CPU both.
+
+    *   Your model must be created in the function ``LightningModule.configure_sharded_model()``.
```
Comment (rohitgr7) [inline docs/source-pytorch/advanced/model_parallel.rst:111]:
```suggestion
    *   Your model must be created using the :meth:`~pytorch_lightning.core.module.LightningModule.configure_sharded_model` method.
```

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 3]
```diff
@@ -55,6 +55,94 @@ Sharding techniques help when model sizes are fairly large; roughly 500M+ parame
 
 ----------
 
+.. _Colossalai:
+
+**************************
+Colossalai
+**************************
+
+ColossalAI implements ZeRO-DP with chunk-based memory management.
```
Comment (rohitgr7) [inline docs/source-pytorch/advanced/model_parallel.rst:64]:
```suggestion
:class:`~pytorch_lightning.strategies.colossalai.ColossalAIStrategy` implements ZeRO-DP with chunk-based memory management.
```

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 4]
```diff
@@ -55,6 +55,94 @@ Sharding techniques help when model sizes are fairly large; roughly 500M+ parame
...
+It is the best choice for users who want to exceptionally enlarge their model size or training batch size.
+
+When using "cuda" option, all parameters are placed in the CUDA memory, no CPU resources will be used during the training.
+It is for users who get plenty of CUDA memory.
+
+The third option, "auto", enables Gemini.
+It monitors the consumption of CUDA memory during the warmup phase and collects CUDA memory usage of all auto-grad operations.
+In later training steps, Gemini manages the data transmission between GPU and CPU automatically according to collected CUDA memory usage information. It is the fastest option when CUDA memory is enough.
+
+Here's an example of changing the placement policy to "cpu".
```
Comment (rohitgr7) [inline docs/source-pytorch/advanced/model_parallel.rst:135]:
```suggestion
Placement policies can help users fully exploit their GPU-CPU heterogeneous memory space for better training efficiency.
There are three options for the placement policy.
They are "cpu", "cuda" and "auto" respectively.

When the placement policy is set to "cpu", all participated parameters will be offloaded into CPU memory immediately at the end of every auto-grad operation.
In this way, "cpu" placement policy uses the least CUDA memory.
It is the best choice for users who want to exceptionally enlarge their model size or training batch size.

When using "cuda" option, all parameters are placed in the CUDA memory, no CPU resources will be used during the training.
It is for users who get plenty of CUDA memory.

The third option, "auto", enables Gemini.
It monitors the consumption of CUDA memory during the warmup phase and collects CUDA memory usage of all auto-grad operations.
In later training steps, Gemini automatically manages the data transmission between GPU and CPU according to collected CUDA memory usage information. It is the fastest option when CUDA memory is enough.

Here's an example of changing the placement policy to "cpu".
```

File: src/pytorch_lightning/strategies/colossalai.py [hunk 5]
```diff
@@ -453,19 +452,6 @@ def reduce(
         tensor = reduce(tensor, dst=0, parallel_mode=ParallelMode.GLOBAL, op=reduce_op)
         return tensor
 
-    def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
```
[the remark this answers is not in the document]
Comment (1SAA) [reply src/pytorch_lightning/strategies/colossalai.py:456]: the method inherited from its supper class is just enough.

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 6]
```diff
@@ -55,6 +55,94 @@ Sharding techniques help when model sizes are fairly large; roughly 500M+ parame
 
 ----------
 
+.. _Colossalai:
+
+**************************
+Colossalai
+**************************
+
```
Comment (awaelchli) [inline docs/source-pytorch/advanced/model_parallel.rst:63]: We could consider adding concise description here as the intro why or when a Lightning user should consider this strategy. Something along the lines of:

> Use when:
> - Training very large, billion parameter models
> - You need to save cost 
> - Train as big of a model you can given a fixed budget of hardware

This is just a dump example of course. But I would consider something like this before going into the technical part of how it works.

File: src/pytorch_lightning/strategies/colossalai.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (1SAA) [reply src/pytorch_lightning/strategies/colossalai.py:456]: ok. I'll fix it.

File: tests/tests_pytorch/strategies/test_colossalai.py [hunk 7]
```diff
@@ -100,7 +100,7 @@ def test_gradient_clip_algorithm_error(tmpdir):
         trainer.fit(model)
 
 
-@RunIf(min_cuda_gpus=1, colossalai=True)
+@RunIf(min_cuda_gpus=1, standalone=True, colossalai=True)
```
Comment (otaj) [inline tests/tests_pytorch/strategies/test_colossalai.py:103]: Why did you make this standalone? This just checks, whether appropriate error is raised, which is happening regardless of standalone.

File: tests/tests_pytorch/strategies/test_colossalai.py [hunk 8]
```diff
@@ -120,7 +120,7 @@ def test_gradient_accumulation_error(tmpdir):
...
```
Comment (otaj) [inline tests/tests_pytorch/strategies/test_colossalai.py:123]: Why did you make this standalone? This just checks, whether appropriate error is raised, which is happening regardless of standalone.

File: src/pytorch_lightning/strategies/colossalai.py [hunk 9]
```diff
@@ -463,8 +463,17 @@ def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
...
-        return broadcast(obj, src=src, parallel_mode=ParallelMode.GLOBAL)
+        if isinstance(obj, Tensor):
+            return broadcast(obj, src=src, parallel_mode=ParallelMode.GLOBAL)
+        else:
+            obj = [obj]
+            local_rank = gpc.get_local_rank(ParallelMode.GLOBAL)
+            if local_rank != src:
+                obj = [None]  # type: ignore[list-item]
+            torch.distributed.broadcast_object_list(obj, src, group=gpc.get_group(ParallelMode.GLOBAL))
+            return obj[0]
```
Comment (otaj) [inline src/pytorch_lightning/strategies/colossalai.py:476]: Adding this code made `mypy` starting to complain quite a bit, please provide typing.

File: tests/tests_pytorch/strategies/test_colossalai.py (hunk 7 above)
Comment (1SAA) [reply tests/tests_pytorch/strategies/test_colossalai.py:103]: There is some deficiency in Colossalai distributed environment initialization. It assumes users using a consistent distributed environment in training. If standalone is False, the testing functions are launched in sequence. Releasing any `ColossalaiStrategy` instance leads to appropriate exist of distributed environment. This causes the next testing function can't initialize correctly. 
It is a problem that should be fixed by Colossalai. But we need to fix it in the next version. Currently, the `standalone` is necessary to be True.

File: docs/source-pytorch/advanced/model_parallel.rst [hunk 10]
```diff
@@ -55,6 +55,100 @@ Sharding techniques help when model sizes are fairly large; roughly 500M+ parame
...
+.. _colossalai:
+
+**********
+Colossalai
+**********
+
+:class:`~pytorch_lightning.strategies.colossalai.ColossalAIStrategy` implements ZeRO-DP with chunk-based memory management.
+With this chunk mechanism, really large models can be trained with restricted computation devices.
+It supprts larger trainable model size and batch size via reducing CUDA memory fragments and CPU memory consumption.
+Also it speeds up this kind of heterogeneous training by fully utilize all kinds of resources.
```
Comment (rohitgr7) [inline docs/source-pytorch/advanced/model_parallel.rst:67]:
```suggestion
:class:`~pytorch_lightning.strategies.colossalai.ColossalAIStrategy` implements ZeRO-DP with chunk-based memory management.
With this chunk mechanism, really large models can be trained with restricted computation devices.
It supports large trainable model size and batch size by reducing CUDA memory fragments and CPU memory consumption.
Also, it speeds up this kind of heterogeneous training by fully utilizing all kinds of resources.
```

**Review by rohitgr7 — APPROVED** [review]
Great! LGTM

File: src/pytorch_lightning/strategies/colossalai.py [hunk 11]
```diff
@@ -463,8 +463,17 @@ def broadcast(self, obj: TBroadcast, src: int = 0) -> TBroadcast:
...
+            from colossalai.core import global_context as gpc
 
-        return broadcast(obj, src=src, parallel_mode=ParallelMode.GLOBAL)
+        if isinstance(obj, Tensor):
+            return broadcast(obj, src=src, parallel_mode=ParallelMode.GLOBAL)
+        else:
+            obj_list = [obj]
+            local_rank = gpc.get_local_rank(ParallelMode.GLOBAL)
+            if local_rank != src:
+                obj_list = [None]  # type: ignore[list-item]
```
Comment (otaj) [inline src/pytorch_lightning/strategies/colossalai.py:474]: Do we need to do this? According to `torch.distributed` [docs](https://pytorch.org/docs/stable/distributed.html#torch.distributed.broadcast_object_list), only data from `src` will be actually broadcast and rest is discarded.

Comment (1SAA) [reply src/pytorch_lightning/strategies/colossalai.py:474]: ok. I'll check it.

**Review by otaj — APPROVED** [review]
